File: Problem1.java
/*
Find object in a grid
approach:
we can do a binary search on row and on column.
for finding on which row object lies, we check for grid[low][0] which is first call, and it is hotter, now calculate
mid and call getResponse api which will give hotter or colder based on previous position which is grid[low][0].
now, if mid is hotter, then we move our low to mid+1. we repeat this until loop breaks and return low which is our target row.
do this for column as well which gives us the target column.
time: O(log m + log n)
space: O(1)
 */
/*
public class Problem1 {
    public int[] findObjectInGrid() {
        int n = maxBound(true),
                m = maxBound(false);
        int objectRow = findRowOrColumn(true, 0, n - 1);
        int objectColumn = findRowOrColumn(false, 0, m - 1);
        return new int[] {objectRow, objectRow};
    }

    private int maxBound(boolean isRow) {
        int low = 0,
                high = Integer.MAX_VALUE;
        while (low <= high) {
            int mid = low + (high - low)/2;
            Response r = getResponse(isRow ? mid : 0, isRow ? 0 : mid);
            if (r.isError()) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }

    private Response getResponse(int i, int i1) {
        return new Response().getResponse(i, i1);
    }

    private int findRowOrColumn(boolean isRow, int start, int end) {
        int low = start,
                high = end;
        while (low <= high) {
            Response r1 = getResponse(isRow ? low : 0, isRow ? 0 : low);
            Response r2 = getResponse(isRow ? high : 0, isRow ? 0 : high);
            int mid = low + (high - low)/2;
            switch (r2.code()) {
                case EXACT:
                    return high;
                case SAME:
                    return mid;
                case HOTTER:
                    low = mid;
                    break;
                case COLDER:
                    high = mid;
                    break;
            }
        }
        return -1;
    }
}

 */

File: problem1.cpp
#include <iostream>
#include <vector>

using namespace std;

enum Response { HOTTER, COLDER, SAME, EXACT };

class Solution {
public:
    Solution() : row(-1), col(-1) {}
    int n;
    int m;
    vector<int> findObject(vector<vector<char>>& grid) {
        n = grid.size();
        m = grid[0].size();
        
        // Traverse the grid to find the object
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (grid[i][j] == 'x') {
                    row = i;
                    col = j;
                    break;
                }
            }
        }
        
        // Binary search to find the row and column of the object
        int lowRow = 0, highRow = n - 1;
        int lowCol = 0, highCol = m - 1;
        
        while (lowRow <= highRow || lowCol <= highCol) {
            int midRow = lowRow + (highRow - lowRow) / 2;
            int midCol = lowCol + (highCol - lowCol) / 2;
            
            Response response = getResponse(midRow, midCol);
            
            if (response == EXACT) {
                return {midRow, midCol};
            } else if (response == HOTTER) {
                lowRow = max(0, midRow - 1);
                highRow = min(n - 1, midRow + 1);
                lowCol = max(0, midCol - 1);
                highCol = min(m - 1, midCol + 1);
            } else if (response == COLDER || response == SAME) {
                if (midRow == lowRow && midCol == lowCol) {
                    lowRow++;
                    lowCol++;
                } else if (midRow == highRow && midCol == highCol) {
                    highRow--;
                    highCol--;
                } else if (midRow == lowRow && midCol == highCol) {
                    lowRow++;
                    highCol--;
                } else if (midRow == highRow && midCol == lowCol) {
                    highRow--;
                    lowCol++;
                } else if (midRow == lowRow) {
                    lowCol = midCol;
                    highCol = midCol;
                } else if (midRow == highRow) {
                    lowCol = midCol;
                    highCol = midCol;
                } else if (midCol == lowCol) {
                    lowRow = midRow;
                    highRow = midRow;
                } else if (midCol == highCol) {
                    lowRow = midRow;
                    highRow = midRow;
                } else if (midRow < row) {
                    lowRow = midRow;
                } else if (midRow > row) {
                    highRow = midRow;
                } else if (midCol < col) {
                    lowCol = midCol;
                } else if (midCol > col) {
                    highCol = midCol;
                }
            }
        }
        
        return {-1, -1}; // Object not found
    }
    
    Response getResponse(int row, int col) {
        if (row < 0 || row >= n || col < 0 || col >= m)
            throw out_of_range("Out of bounds");
        
        if (row == this->row && col == this->col)
            return EXACT;
        
        int distance = abs(this->row - row) + abs(this->col - col);
        int prevDistance = abs(this->row - prevRow) + abs(this->col - prevCol);
        
        if (distance < prevDistance)
            return HOTTER;
        else if (distance > prevDistance)
            return COLDER;
        else
            return SAME;
    }
    
private:
    int row, col;
    int prevRow, prevCol;
};

int main() {
    Solution solution;

    vector<vector<char>> grid1 = {
        {'o', 'o', 'o'},
        {'o', 'o', 'o'},
        {'x', 'o', 'o'}
    };
    vector<int> result1 = solution.findObject(grid1);
    cout << "Object found at: [" << result1[0] << ", " << result1[1] << "]" << endl;

    vector<vector<char>> grid2 = {
        {'o', 'o', 'o', 'o', 'o'},
        {'o', 'o', 'o', 'o', 'o'},
        {'o', 'o', 'o', 'o', 'o'},
        {'o', 'o', 'o', 'o', 'o'},
        {'o', 'o', 'o', 'x', 'o'},
        {'o', 'o', 'o', 'o', 'o'}
    };
    vector<int> result2 = solution.findObject(grid2);
    cout << "Object found at: [" << result2[0] << ", " << result2[1] << "]" << endl;

    return 0;
}
